com.aobuchow.themes.spectrum.preferences/src/com/aobuchow/themes/spectrum/preferences/ColorManager.java

```diff
@@ -1,34 +1,63 @@
 package com.aobuchow.themes.spectrum.preferences;
 
+import org.eclipse.e4.core.contexts.IEclipseContext;
+import org.eclipse.e4.ui.css.swt.theme.IThemeEngine;
+import org.eclipse.e4.ui.model.application.MApplication;
 import org.eclipse.jface.resource.ColorRegistry;
 import org.eclipse.swt.SWT;
 import org.eclipse.swt.custom.StyleRange;
 import org.eclipse.swt.custom.StyledText;
 import org.eclipse.swt.graphics.Color;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.internal.util.PrefUtil;
+
+import com.aobuchow.themes.spectrum.preferences.ColorHSL.BOUND_BEHAVIOR;
+import com.aobuchow.themes.spectrum.preferences.ColorHSL.HSL_PROPERTY;
+import com.aobuchow.themes.spectrum.preferences.ColorHSL.MODIFICATION;
 
 public class ColorManager {
+	private ColorRegistry colorRegistry;
+	private Color accentColor;
+	private Color baseColor;
+	private Color backgroundColor;
+	private IThemeEngine engine;
```

Comment:
``ThemeEngine`` is no API (anymore?) in 2020-09.
From my experience, using internal APIs works for the moment but makes migration expensive when you want to use your bundle w/ a newer Eclipse version.

Reply:
Yes agreed, I shouldn't be using any internal API's so I'd like to remove all of theses instances of discouraged access. I do still need to find alternatives to all the internal API's I'm using however.

Reply:
From what it seems, ITheme engine is not (and maybe never was?) API. However, it's still mentioned [here](https://www.vogella.com/tutorials/Eclipse4CSS/article.html#styling-using-themes), although DI is used to retrieve the IThemeEngine instance.

IIRC none of Eclipse's CSS engine related stuff is actual API? See https://bugs.eclipse.org/bugs/show_bug.cgi?id=434201#c23

So I'm not sure if there's any external API that could be used instead of IThemeEngine.

Reply:
> From what it seems, ITheme engine is not (and maybe never was?) API. However, it's still mentioned [here](https://www.vogella.com/tutorials/Eclipse4CSS/article.html#styling-using-themes), although DI is used to retrieve the IThemeEngine instance.
> 
> IIRC none of Eclipse's CSS engine related stuff is actual API? See https://bugs.eclipse.org/bugs/show_bug.cgi?id=434201#c23
> 
> So I'm not sure if there's any external API that could be used instead of IThemeEngine.

I don't know enough about theming (and its APIs). Maybe, an upstream issue could help there. You know best.
If it's not dangerous (i.e. if there's no other API, other contributors will also use this - which makes it defacto-API), it could stay the way it is, I guess :)

Reply:
No worries. IMO they should probably just make this internal API into external API, as you mentioned, it's basically already the defacto-API :P